File: backend/server/src/main/kotlin/nl/knaw/huc/dexter/api/Source.kt
package nl.knaw.huc.dexter.api

import java.time.LocalDate
import java.time.LocalDateTime
import java.util.*

data class FormSource(
    val externalRef: String? = null,
    val externalId: String? = null,
    val title: String,
    val description: String? = null,
    val rights: String? = null,
    val access: AccessType? = null,
    val creator: String? = null,
    val location: String? = null,
    val earliest: LocalDate? = null,
    val latest: LocalDate? = null,
    val notes: String? = null,
    val ethics: String? = null,
)

data class ResultSource(
    val id: UUID,
    val externalRef: String? = null,
    val externalId: String? = null,
    val title: String,
    val description: String? = null,
    val rights: String? = null,
    val access: AccessType? = null,
    val creator: String? = null,
    val location: String? = null,
    val earliest: LocalDate? = null,
    val latest: LocalDate? = null,
    val notes: String? = null,
    val ethics: String? = null,
    val createdBy: UUID,
    val createdAt: LocalDateTime,
    val updatedAt: LocalDateTime
)

/**
 * With associated child resources
 */
data class ResultSourceWithChildIds (
    val id: UUID,
    val externalRef: String? = null,
    val externalId: String? = null,
    val title: String,
    val description: String? = null,
    val rights: String? = null,
    val access: AccessType? = null,
    val creator: String? = null,
    val location: String? = null,
    val earliest: LocalDate? = null,
    val latest: LocalDate? = null,
    val notes: String? = null,
    val ethics: String? = null,
    val createdBy: UUID,
    val createdAt: LocalDateTime,
    val updatedAt: LocalDateTime,

    val references: List<UUID>,
    val corpora: List<UUID>,
    val languages: List<String>,
    val media: List<UUID>,
    val metadataValues: List<UUID>,
    val tags: List<Int>
)

fun ResultSource.toResultSourceWithChildIds(
    references: List<UUID>,
    corpora: List<UUID>,
    languages: List<String>,
    media: List<UUID>,
    metadataValues: List<UUID>,
    tags: List<Int>
) = ResultSourceWithChildIds(
    id = id,
    externalRef = externalRef,
    externalId = externalId,
    title = title,
    description = description,
    rights = rights,
    access = access,
    creator = creator,
    location = location,
    earliest = earliest,
    latest = latest,
    notes = notes,
    ethics = ethics,
    createdBy = createdBy,
    createdAt = createdAt,
    updatedAt = updatedAt,

    references = references,
    corpora = corpora,
    languages = languages,
    media = media,
    metadataValues = metadataValues,
    tags = tags
)

File: frontend/src/model/Source.ts
import { ResultCorpus } from './Corpus';
import { Any } from '../components/common/Any';
import { ResultTag } from './Tag';
import { FormMetadataValue, MetadataValue } from './Metadata';
import { ResultMedia } from './Media';
import { ResultLanguage } from './Language';
import { ResultReference } from './Reference';
import { UUID, WithId } from './Id';
import { Access } from './Access';
import { LocalDate, LocalDateTime } from './Date';
import _ from 'lodash';

export type FormSource = {
  title: string;
  description?: string;
  rights?: string;
  access?: Access;
  creator?: string;
  externalRef?: string;
  externalId?: string;
  location?: string;
  earliest?: LocalDate;
  latest?: LocalDate;
  notes?: string;
  ethics?: string;
};
/**
 * Source result as send by server
 */
export type ResultSource = FormSource &
  WithId & {
    createdBy: UUID;
    createdAt: LocalDateTime;
    updatedAt: LocalDateTime;
  };
export type ResultSourceWithChildIds = ResultSource & {
  references: UUID[];
  corpora: UUID[];
  languages: string[];
  media: UUID[];
  metadataValues: UUID[];
  tags: number[];
};

export function toResultSourceWithChildren(
  result: ResultSource,
): ResultSourceWithChildIds {
  return {
    ...result,
    references: [],
    corpora: [],
    languages: [],
    media: [],
    metadataValues: [],
    tags: [],
  };
}

/**
 * Source including all child resources
 */
export type Source = ResultSource & {
  references: ResultReference[];
  corpora: ResultCorpus[];
  languages: ResultLanguage[];
  media: ResultMedia[];
  metadataValues: MetadataValue[];
  tags: ResultTag[];
};

export function isSource(toTest: Any): toTest is Source {
  if (!toTest) {
    return false;
  }
  return (
    !_.isUndefined((toTest as Source)?.description) &&
    !_.isUndefined((toTest as Source)?.media) &&
    !_.isUndefined((toTest as Source)?.references)
  );
}

export type SubmitFormSource = Omit<Source, 'id' | 'metadataValues'> & {
  metadataValues: FormMetadataValue[];
};
